File: benchmarks/ocaml/mandelbrot.ml
(* Mandelbrot set calculation benchmark *)

(* Size constants *)
let width = 800
let height = 800
let max_iterations = 100

(* Calculate the Mandelbrot set *)
let calculate_mandelbrot max_iterations =
  Array.init height (fun y ->
    Array.init width (fun x ->
      let cx = (float_of_int x -. float_of_int width /. 2.0) *. 4.0 /. float_of_int width in
      let cy = (float_of_int y -. float_of_int height /. 2.0) *. 4.0 /. float_of_int height in
      
      let rec iterate zx zy iteration =
        if zx *. zx +. zy *. zy > 4.0 || iteration >= max_iterations then
          iteration
        else
          let temp = zx *. zx -. zy *. zy +. cx in
          iterate temp (2.0 *. zx *. zy +. cy) (iteration + 1)
      in
      
      iterate 0.0 0.0 0
    )
  )

(* Count non-black pixels in the result *)
let count_non_black result max_iterations =
  Array.fold_left (fun acc row ->
    acc + Array.fold_left (fun acc' pixel ->
      if pixel < max_iterations then acc' + 1 else acc'
    ) 0 row
  ) 0 result

let () =
  let start_time = Unix.gettimeofday () in
  
  let result = calculate_mandelbrot max_iterations in
  let count = count_non_black result max_iterations in
  
  Printf.printf "Mandelbrot set calculation finished.\n";
  Printf.printf "Image size: %d x %d\n" width height;
  Printf.printf "Maximum iterations: %d\n" max_iterations;
  Printf.printf "Non-black pixels: %d\n" count;
  
  let elapsed = int_of_float ((Unix.gettimeofday () -. start_time) *. 1000.0) in
  Printf.printf "Time taken: %d ms\n" elapsed;
  
  (* Approximate memory usage *)
  let memory_usage = Gc.stat () in
  Printf.printf "Memory used: %d KB\n" (memory_usage.minor_words |> int_of_float |> (fun x -> x / 1024))

File: benchmarks/rust/mandelbrot.rs
// Mandelbrot set calculation benchmark

use std::time::Instant;

// Size constants
const WIDTH: usize = 800;
const HEIGHT: usize = 800;
const MAX_ITERATIONS: usize = 100;

// Calculate the Mandelbrot set
fn calculate_mandelbrot(max_iterations: usize) -> Vec<Vec<usize>> {
    let mut result = vec![vec![0; WIDTH]; HEIGHT];
    
    for y in 0..HEIGHT {
        for x in 0..WIDTH {
            let cx = (x as f64 - WIDTH as f64 / 2.0) * 4.0 / WIDTH as f64;
            let cy = (y as f64 - HEIGHT as f64 / 2.0) * 4.0 / HEIGHT as f64;
            
            let mut zx = 0.0;
            let mut zy = 0.0;
            let mut iteration = 0;
            
            while zx*zx + zy*zy <= 4.0 && iteration < max_iterations {
                let temp = zx*zx - zy*zy + cx;
                zy = 2.0 * zx * zy + cy;
                zx = temp;
                iteration += 1;
            }
            
            result[y][x] = iteration;
        }
    }
    
    result
}

// Count non-black pixels in the result
fn count_non_black(result: &Vec<Vec<usize>>, max_iterations: usize) -> usize {
    let mut count = 0;
    
    for y in 0..HEIGHT {
        for x in 0..WIDTH {
            if result[y][x] < max_iterations {
                count += 1;
            }
        }
    }
    
    count
}

fn main() {
    let start_time = Instant::now();
    
    let result = calculate_mandelbrot(MAX_ITERATIONS);
    let count = count_non_black(&result, MAX_ITERATIONS);
    
    println!("Mandelbrot set calculation finished.");
    println!("Image size: {} x {}", WIDTH, HEIGHT);
    println!("Maximum iterations: {}", MAX_ITERATIONS);
    println!("Non-black pixels: {}", count);
    
    let elapsed = start_time.elapsed();
    println!("Time taken: {} ms", elapsed.as_millis());
    
    // Get approximate memory usage
    let memory_usage = std::mem::size_of::<Vec<Vec<usize>>>() + (HEIGHT * WIDTH * std::mem::size_of::<usize>());
    println!("Memory used: {} KB", memory_usage / 1024);
}